### DISCUSSION TITLE
Parallelization on CPU only

### DISCUSSION THREAD

**Discussion by npiroozan** [deepmodeling/deepmd-kit#2544] [Q&A]
Hello,

I hope you are doing well.

If possible, I have a question regarding the way the model works with LAMMPS on CPU only.

When I run the trained model (for Silicon) with LAMMPS, I do not see any usage of neighbor lists or OMP parallelization in the simulation.  It seems like it is only MPI parallelized.  I think I'm doing something wrong, but would you be so kind as to review the Training Script and the LAMMPS Script?  Thank you very much.

Training Script:

{
    "model": {
        "type_map":     ["Si"],
        "data_stat_nbatch": 10,

        "descriptor" :{
            "type":             "se_e2_a",
            "sel":              [46],
            "rcut_smth":        0.50,
            "rcut":             6.0,
            "resnet_dt":        false,
            "precision":        "bfloat16"
        },

        "fitting_net" : {
            "type":             "ener",
            "neuron":           [240, 240, 240],
            "activation_function": "tanh",
            "precision":        "bfloat16",
            "resnet_dt":        true,
            "seed":             1
        }
    },

    "learning_rate" :{
        "scale_by_worker": "none",
        "type":         "exp",
        "decay_steps":  472,
        "start_lr":     0.024905919858219574,
        "stop_lr":      5.484673842346183e-9
    },

    "loss" :{
        "type":         "ener",
        "start_pref_e": 0.02,
        "limit_pref_e": 1,
        "start_pref_f": 1000,
        "limit_pref_f": 1,
        "start_pref_v": 0,
        "limit_pref_v": 0
    },

    "training" : {
        "training_data": {
            "systems":          ["/localdisk/npirooza/Input_Data/Silicon_Original/data_0/", "/localdisk/npirooza/Input_Data/Silicon_Original/data_1/", "/localdisk/npirooza/Input_Data/Silicon_Original/data_2/", "/localdisk/npirooza/Input_Data/Silicon_Original/data_3/", "/localdisk/npirooza/Input_Data/Silicon_Original/data_4", "/localdisk/npirooza/Input_Data/Silicon_Original/data_5", "/localdisk/npirooza/Input_Data/Silicon_Original/data_6", "/localdisk/npirooza/Input_Data/Silicon_Original/data_7", "/localdisk/npirooza/Input_Data/Silicon_Original/data_8"],
            "batch_size":       32
        },

        "validation_data":{
            "systems":          ["/localdisk/npirooza/Input_Data/Silicon_Original/data_9"],
            "batch_size":       32
        },

        "mixed_precision": {
                "output_prec": "float32",
                "compute_prec": "bfloat16"
        },

        "numb_steps":   32000,
        "seed":         1,
        "disp_file":    "lcurve.out",
        "disp_freq":    100,
        "save_freq":    1000000,
        "numb_test":    10,
        "save_ckpt":    "model.ckpt",
        "disp_training": true,
        "time_training": true,
        "tensorboard":  false,
        "tensorboard_log_dir": "log",
        "tensorboard_freq": 100,
        "enable_profiler": false,
        "profiling": false,
        "profiling_file": "timeline.json"
    }
}

LAMMPS Input Script:

## Initialize

units           metal
dimension       3
boundary        p p p
atom_style      atomic
#plugin          load libdeepmd_lmp.so
package         intel 0 omp 1 mode double

## System

lattice         diamond 5.4307 orient x 1 0 0 orient y 0 1 0 orient z 0 0 1
region          sim block 0 7 0 8 0 9
create_box      1 sim
create_atoms    1 region sim
mass            1 28.0855

## Force Field Selection

pair_style      deepmd graph-compress-high.pb
pair_coeff      * *
timestep        0.002

## Neighbor List
neighbor        2.5 bin
neigh_modify    delay 10 check yes

# Define Outputs

thermo          100
thermo_style    custom step temp pe ke etotal
log             log.energy

# Dynamics

#### Minimize Energy ####

min_style       cg
minimize        1.0e-8 1.0e-8 5000 10000
reset_timestep  0

#### Thermalization ####

velocity        all create 300.0 65897 mom yes rot yes

fix             1 all langevin 300.0 300.0 1 32456
fix             2 all nve
run             25000
unfix           1
unfix           2

fix            solid all nvt temp 375.0 375.0 2.5 drag 2.0
run            50000
unfix          solid

Output LAMMPS:

MPI task timing breakdown:
Section |  min time  |  avg time  |  max time  |%varavg| %total
---------------------------------------------------------------
Pair    | 29.41      | 39.333     | 51.029     |  63.7 | 68.44
Neigh   | 0          | 0          | 0          |   0.0 |  0.00
Comm    | 4.116      | 15.128     | 25.341     | 103.7 | 26.32
Output  | 0.018891   | 0.019464   | 0.025107   |   0.3 |  0.03
Modify  | 1.1884     | 2.4839     | 4.4482     |  38.7 |  4.32
Other   |            | 0.503      |            |       |  0.88

Thank you very much!

**Comment by njzjz** [comment]
>  I do not see any usage of neighbor lists or OMP parallelization in the simulation. It seems like it is only MPI parallelized.

Could you explain how you saw that?

**Reply by npiroozan** [reply]
Certainly.  When I set omp num threads to a value, N, and mpi ranks (on a single node) to M, the output from LAMMPS gives:

M mpi ranks with 1 OMP threads for 16000 atoms.

Whatever I use for the setting of OMP, LAMMPS will accept only OMP=1.

I have seen these behavior for ReaxFF in the past, until the development of reax/omp.

Thank you!

**Reply by njzjz** [reply]
The OMP in the deepmd-kit is only controlled by the environment variable `OMP_NUM_THREADS`. Any configurations except  `OMP_NUM_THREADS` will not be passed to deepmd-kit.

**Reply by npiroozan** [reply]
Apologies, I used an abbreviation.

I set OMP_NUM_THREADS=4

For example, but LAMMPS did not recognize it.

**Reply by njzjz** [reply]
DeePMD-kit doesn't fetch openmp configurations from LAMMPS, so if LAMMPS has something wrong with openmp, it doesn't affect DeePMD-kit.

**Reply by npiroozan** [reply]
I understand, and thank you for the clarification.

If possible, allow me to please go through the process with which I build Deepmd-kit and LAMMPS.

1.  Within the conda environment,
Build Tensorflow 2.9
Horovod 0.5
DeePMD 2.1.5
Deepmodeling package lammps
Additional packages as required

2.  Import LAMMPS-23-2022 Stable release

3.  Install the Tensorflow 2.9 C++ API
4.  Build DeePMD C++ API

5.  Build LAMMPS using the Built-in method (make)

6.  Train the model successfully, freeze, compress



7.  Run the DeePMD+LAMMPS simulation with the following run script:


nodes=1
ppn=16
omp=4
intra=1
inter=1
kmp_blk=1

export OMP_NUM_THREADS=4

verbose="MKLDNN_VERBOSE 1 -genv MKL_VERBOSE 1 -genv DNNL_VERBOSE 2"
kmp="OMP_NUM_THREADS ${omp} -genv KMP_BLOCKTIME=${kmp_blk} "
tf_intra="TF_INTRA_OP_PARALLELISM_THREADS ${intra}"
tf_inter="TF_INTER_OP_PARALLELISM_THREADS ${inter}"
hvd_ccl_ops="HOROVOD_CPU_OPERATIONS=CCL"
hvd_cache="HOROVOD_CCL_CACHE=1"
tf_onednn_opts="TF_ENABLE_ONEDNN_OPTS=1"

export DP_INTERFACE_PREC=high
export LAMMPS_PLUGIN_PATH=/pathway/deepmd-kit-deepmd29_py309-bf16/source/lib/


mpirun -hostfile ${hostfile} -genvall -bootstrap ssh -n $[nodes*ppn] -ppn ${ppn} \
     -genv ${kmp} \
     -genv ${tf_onednn_opts} \
     -genv ${hvd_ccl_ops} \
     -genv ${hvd_cache} \
     -genv DP_AUTO_PARALLELIZATION 1 \
     -genv DP_INTERFACE_PREC high \
     -genv ${tf_intra}\
     -genv ${tf_inter}\
      lmp-in in.silicon




Output:


Per MPI rank memory allocation (min/avg/max) = 3.127 | 3.127 | 3.127 Mbytes
   Step          Temp          PotEng         KinEng         TotEng
         0   0             -430966.3       0             -430966.3
       100   0.00036324967 -430966.29      0.00018927038 -430966.29
       200   0.00039487705 -430966.29      0.00020574976 -430966.29
       300   0.0005445072  -430966.29      0.00028371419 -430966.29
       400   0.00071458047 -430966.29      0.00037233047 -430966.29
       500   0.00086712436 -430966.29      0.0004518131  -430966.28
       600   0.0012106785  -430966.28      0.00063082117 -430966.28
       700   0.0016087958  -430966.29      0.00083825928 -430966.28
       800   0.0019382942  -430966.28      0.0010099436  -430966.28
       900   0.0023331417  -430966.28      0.001215678   -430966.28
      1000   0.0029632348  -430966.28      0.0015439865  -430966.28
      1100   0.0040777137  -430966.28      0.0021246831  -430966.28
      1200   0.0054590698  -430966.28      0.0028444355  -430966.28
      1300   0.0075706045  -430966.28      0.0039446456  -430966.28
      1400   0.010148215   -430966.28      0.0052877034  -430966.28
      1500   0.012751618   -430966.28      0.0066442008  -430966.27
      1600   0.016188529   -430966.28      0.0084349951  -430966.27
      1700   0.021823852   -430966.27      0.011371267   -430966.26
      1800   0.02917592    -430966.27      0.015202045   -430966.26
      1900   0.041371605   -430966.27      0.021556577   -430966.24
      2000   0.05896553    -430966.26      0.03072385    -430966.23
Loop time of 41.601 on 16 procs for 2000 steps with 4032 atoms

Performance: 8.307 ns/day, 2.889 hours/ns, 48.076 timesteps/s
365.4% CPU use with 16 MPI tasks x 1 OpenMP threads

MPI task timing breakdown:
Section |  min time  |  avg time  |  max time  |%varavg| %total
---------------------------------------------------------------
Pair    | 33.135     | 37.443     | 41.404     |  40.3 | 90.01
Neigh   | 0          | 0          | 0          |   0.0 |  0.00
Comm    | 0.15689    | 3.6298     | 8.42       | 135.9 |  8.73
Output  | 0.00045482 | 0.00048028 | 0.00063355 |   0.0 |  0.00
Modify  | 0.025565   | 0.5144     | 1.9724     | 117.3 |  1.24
Other   |            | 0.01328    |            |       |  0.03



This line in particular is of concern:

365.4% CPU use with 16 MPI tasks x 1 OpenMP threads

OpenMP Threads has not been recognized as being 4.  Is there anything wrong with my implementation from what you can see?


Thank you very much for your time.

**Reply by njzjz** [reply]
Probably Your LAMMPS didn't enable the OpenMP flag.
https://github.com/lammps/lammps/blob/ed03d04a70551a4ffbf9116e714eb51c466c63a5/src/comm.cpp#L92-L102
The codes show that LAMMPS recognizes the threads only when `_OPENMP` is defined.

However, the CPU usage is acceptable (365.4%), close to 4.

**Reply by npiroozan** [reply]
Thank you, I think it clarifies the situation.

**Reply by npiroozan** [reply]
One last item, I apologize:

Why would neighbor lists not be used?  It was applied properly in the model during training under "sel".

**Reply by njzjz** [reply]
Usually the neighbor information is printed in the beginning. Could you find it?

**Reply by npiroozan** [reply]
Sorry for the delayed response.  I believe you are referring to this?:

Neighbor list info ...
  update every 1 steps, delay 10 steps, check yes
  max neighbors/atom: 2000, page size: 100000
  master list distance cutoff = 8
  ghost atom cutoff = 8
  binsize = 4, bins = 10 11 13
  1 neighbor lists, perpetual/occasional/extra = 1 0 0

**Reply by njzjz** [reply]
Thanks, then I think the neighbor list is computed correctly.

**Reply by npiroozan** [reply]
Seems so and is properly configured in the model (sel is not too small or too large).

**Reply by npiroozan** [reply]
Would it be fair to say that DeePMD Inference is highly optimized for CPUs?

Specifically if the trained model is well parallelized.

Thank you very much.

**Reply by npiroozan** [reply]
The reason that I say this is due the following:

I am running on a 56 core CPU, and performance decreases as OPENMP_NUM_THREADS increases.

Single Node
4032 atoms

ppn=16, OPENMP_NUM_THREADS=1        Throughput for NVT Thermalization = 82 timesteps/sec
ppn=16, OPENMP_NUM_THREADS=2        Throughput for NVT Thermalization = 49 timesteps/sec
ppn=16, OPENMP_NUM_THREADS=4        Throughput for NVT Thermalization = 27 timesteps/sec
ppn=16, OPENMP_NUM_THREADS=8        Throughput for NVT Thermalization = 15 timesteps/sec

This seems a bit strange to me.

Indeed, the best possible performance on my CPU is the following:

ppn=112, OPENMP_NUM_THREADS=1          Throughput for NVT Thermalization = 380 timesteps/sec

I am surprised that an HPC code uses so many MPI ranks.  The issue of shared memory, and communication time between processes becomes a big factor.

Why is the code behaving this way?

**Reply by njzjz** [reply]
I need to understand how to have 112 threads on 56 cores - don't these threads compete for resources with each other?

**Reply by njzjz** [reply]
Also, 4032 atoms might be too few for 112 threads.

**Reply by npiroozan** [reply]
We have 2 sockets with 56 cores on each. Each socket is split in two quadrants. Each quadrant has hbm and ddr memory across multiple numa bands.  A total of 532 GB memory.

Same behavior exists if I increase the number of atoms significantly.

**Reply by npiroozan** [reply]
The way that I reached that configuration was based on maximizing the throughput.  What would you recommend to do?

**Reply by njzjz** [reply]
Thanks for clarifying. Could you try `KMP_BLOCKTIME=0` (I see you set it to 1 in the previous comment)?

**Reply by npiroozan** [reply]
I'll try it now and let you know if there is any change in performance.

**Reply by njzjz** [reply]
When you built tensorflow, was oneDNN enabled?

If not, tensorflow only uses Eigen for parallelism. OMP_NUM_THREADS does not work for that.

**Reply by npiroozan** [reply]
C++ API of Tensorflow is currently using Eigen based on Vtune analysis. I thought I built it with onednn...

**Reply by npiroozan** [reply]
Python API of tensorflow is with oneDNN and mkl support. Running the major ops as mkl_MATMUL, etc.

C++ api is eigen now. I'm sorry to ask but what is the procedure to build C++ Tensorflow with mkl and onednn support?

**Reply by njzjz** [reply]
Append `--config=mkl` to `bazel build`.

**Reply by npiroozan** [reply]
When the Python api is used for DeePMD and LAMMPS simulation, openmp_num_threads is also behaving in the same way as my problem.  Is this expected?

**Reply by njzjz** [reply]
According to [the TensorFlow blog](https://blog.tensorflow.org/2022/05/whats-new-in-tensorflow-29.html), oneDNN needs to be enabled by `TF_ENABLE_ONEDNN_OPTS=1`. Is this helpful?

**Reply by npiroozan** [reply]
Thank you very much. Will reinstall Tensorflow C++ api with mkl to ensure we are no longer using Eigen.

This, along with your additional suggestion I believe will resolve the issue of OMP parallelism.

Is that correct?

**Reply by npiroozan** [reply]
OpenMP parallelism is working and fixed the lammps flag issue.

There seems to be a large increase in cache memory stalls when multiple threads/core are used. 

This is what resulted in a decrease in throughput. Have you seen this issue with CPUs?

**Reply by njzjz** [reply]
LAMMPS uses a clever distributed-memory MPI parallelism. Each processor stores information for the subset of atoms within its subdomain. So, there is only a little communication between processors. See [the LAMMPS paper](https://doi.org/10.1016/j.cpc.2021.108171) for details.
OpenMP parallelism is within the TensorFlow OPs. In principle, OpenMP may need more communication to exchange variables.

**Reply by npiroozan** [reply]
Thank you very much. Definitely there seems to be areas of improvement for performance on CPUs.  Perhaps we can discuss further offline?
